README
# Добро пожаловать в репозиторий DHT!

DHT - это программный код, позволяющий считывать данные с датчиков трех типов,
DHT 11, DHT 22, DHT 2302.

Данный код написан для архитектуры arm с использованием чипов A10, A20 и др.
Обращение происходит не через gpio_sunxi, а напрямую, изменяя состояние регистров.

Настройка:

В файле src/main.c нужно изменить порт GPIO на свой.

По умолчанию принято:
PG3  SUNXI_GPG(3)
int dhtpin = PG3

Измените порт на свой!
PXY  SUNXI_GPX(Y)
int dhtpin = PXY
X - буквенная часть A, B, C, D, E, F, I.
Y - номер пина.
XY - задаются в .fex файле вашего микрокомпьютера.

Для компиляции можно использовать 2 варианта:

1) Стандартным методом:
./configure 
./make
2) Открыть проект файла в IDE Anjuta и собрать проект.

Синтаксис программы:
dht [model]
dht [11|22|2302]


ChangeLog
- Исправлена проблема получения данных с датчика при загруженности компьютера.
(Данные не считывались если микрокомпьютер был загружен более чем на 30%)
- Исправлен баг вхождения программы в бесконечный цикл если опрос датчика проходил слишком часто.
- Исправлена выдача результата, если данные считались не полностью. Теперь, если данные считались не полностью,
программа повторит попытку чтения данных повторно.


config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.ac by autoheader.  */

/* always defined to indicate that i18n is enabled */
#define ENABLE_NLS 1

/* GETTEXT package name */
#define GETTEXT_PACKAGE "dht"

/* Define to 1 if you have the `bind_textdomain_codeset' function. */
#define HAVE_BIND_TEXTDOMAIN_CODESET 1

/* Define to 1 if you have the `dcgettext' function. */
#define HAVE_DCGETTEXT 1

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define if the GNU gettext() function is already present or preinstalled. */
#define HAVE_GETTEXT 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define if your <locale.h> file defines LC_MESSAGES. */
#define HAVE_LC_MESSAGES 1

/* Define to 1 if you have the <locale.h> header file. */
#define HAVE_LOCALE_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to the sub-directory where libtool stores uninstalled libraries. */
#define LT_OBJDIR ".libs/"

/* Name of package */
#define PACKAGE "dht"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME "dht"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "dht 0.1"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "dht"

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION "0.1"

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "0.1"


src/gpio_lib.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*-  */
/*
 * gpio_lib.c
 * Copyright (C) 2015 Stefan Mavrodiev <support@olimex.com>
 *
 * DHT is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * DHT is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/select.h>
#include <pthread.h>
#include <unistd.h>
#include <sched.h>
#include "gpio_lib.h"
unsigned int SUNXI_PIO_BASE = 0;
static volatile long int *gpio_map = NULL;
int sunxi_gpio_init(void) {
int fd;
unsigned int addr_start, addr_offset;
unsigned int PageSize, PageMask;
fd = open("/dev/mem", O_RDWR);
if(fd < 0) {
return SETUP_DEVMEM_FAIL;
}
PageSize = sysconf(_SC_PAGESIZE);
PageMask = (~(PageSize-1));
addr_start = SW_PORTC_IO_BASE & PageMask;
addr_offset = SW_PORTC_IO_BASE & ~PageMask;
gpio_map = (void *)mmap(0, PageSize*2, PROT_READ|PROT_WRITE, MAP_SHARED, fd, addr_start);
if(gpio_map == MAP_FAILED) {
return SETUP_MMAP_FAIL;
}
SUNXI_PIO_BASE = (unsigned int)gpio_map;
SUNXI_PIO_BASE += addr_offset;
close(fd);
return SETUP_OK;
}
int sunxi_gpio_set_cfgpin(unsigned int pin, unsigned int val) {
unsigned int cfg;
unsigned int bank = GPIO_BANK(pin);
unsigned int index = GPIO_CFG_INDEX(pin);
unsigned int offset = GPIO_CFG_OFFSET(pin);
if(SUNXI_PIO_BASE == 0) {
return -1;
}
struct sunxi_gpio *pio =
&((struct sunxi_gpio_reg *)SUNXI_PIO_BASE)->gpio_bank[bank];
cfg = *(&pio->cfg[0] + index);
cfg &= ~(0xf << offset);
cfg |= val << offset;
*(&pio->cfg[0] + index) = cfg;
return 0;
}
int sunxi_gpio_get_cfgpin(unsigned int pin) {
unsigned int cfg;
unsigned int bank = GPIO_BANK(pin);
unsigned int index = GPIO_CFG_INDEX(pin);
unsigned int offset = GPIO_CFG_OFFSET(pin);
if(SUNXI_PIO_BASE == 0)
{
return -1;
}
struct sunxi_gpio *pio = &((struct sunxi_gpio_reg *)SUNXI_PIO_BASE)->gpio_bank[bank];
cfg = *(&pio->cfg[0] + index);
cfg >>= offset;
return (cfg & 0xf);
}
int sunxi_gpio_output(unsigned int pin, unsigned int val) {
unsigned int bank = GPIO_BANK(pin);
unsigned int num = GPIO_NUM(pin);
if(SUNXI_PIO_BASE == 0)
{
return -1;
}
struct sunxi_gpio *pio =&((struct sunxi_gpio_reg *)SUNXI_PIO_BASE)->gpio_bank[bank];
if(val)
*(&pio->dat) |= 1 << num;
else
*(&pio->dat) &= ~(1 << num);
return 0;
}
int sunxi_gpio_input(unsigned int pin) {
unsigned int dat;
unsigned int bank = GPIO_BANK(pin);
unsigned int num = GPIO_NUM(pin);
if(SUNXI_PIO_BASE == 0)
{
return -1;
}
struct sunxi_gpio *pio =&((struct sunxi_gpio_reg *)SUNXI_PIO_BASE)->gpio_bank[bank];
dat = *(&pio->dat);
dat >>= num;
return (dat & 0x1);
}
void sunxi_gpio_cleanup(void)
{
unsigned int PageSize;
if (gpio_map == NULL)
return;
PageSize = sysconf(_SC_PAGESIZE);
munmap((void*)gpio_map, PageSize*2);
}


src/gpio_lib.h
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*-  */
/*
 * gpio_lib.h
 * Copyright (C) 2015 Stefan Mavrodiev <support@olimex.com>
 *
 * DHT is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * DHT is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _GPIO_LIB_H_
#define _GPIO_LIB_H_
#define SW_PORTC_IO_BASE 0x01c20800
#define SUNXI_GPIO_A	0
#define SUNXI_GPIO_B	1
#define SUNXI_GPIO_C	2
#define SUNXI_GPIO_D	3
#define SUNXI_GPIO_E	4
#define SUNXI_GPIO_F	5
#define SUNXI_GPIO_G	6
#define SUNXI_GPIO_H	7
#define SUNXI_GPIO_I	8
#define SETUP_OK 0
#define SETUP_DEVMEM_FAIL 1
#define SETUP_MALLOC_FAIL 2
#define SETUP_MMAP_FAIL 3
#define HIGH 1
#define LOW 0
#define INPUT 0
#define OUTPUT 1
#define PER 2
struct sunxi_gpio {
unsigned int cfg[4];
unsigned int dat;
unsigned int drv[2];
unsigned int pull[2];
};
/* gpio interrupt control */
struct sunxi_gpio_int {
unsigned int cfg[3];
unsigned int ctl;
unsigned int sta;
unsigned int deb;
};
struct sunxi_gpio_reg {
struct sunxi_gpio gpio_bank[9];
unsigned char res[0xbc];
struct sunxi_gpio_int gpio_int;
};
#define GPIO_BANK(pin) ((pin) >> 5)
#define GPIO_NUM(pin) ((pin) & 0x1F)
#define GPIO_CFG_INDEX(pin) (((pin) & 0x1F) >> 3)
#define GPIO_CFG_OFFSET(pin) ((((pin) & 0x1F) & 0x7) << 2)
/* GPIO bank sizes */
#define SUNXI_GPIO_A_NR	(32)
#define SUNXI_GPIO_B_NR	(32)
#define SUNXI_GPIO_C_NR	(32)
#define SUNXI_GPIO_D_NR	(32)
#define SUNXI_GPIO_E_NR	(32)
#define SUNXI_GPIO_F_NR	(32)
#define SUNXI_GPIO_G_NR	(32)
#define SUNXI_GPIO_H_NR	(32)
#define SUNXI_GPIO_I_NR	(32)
#define SUNXI_GPIO_NEXT(__gpio) ((__gpio##_START)+(__gpio##_NR)+0)
enum sunxi_gpio_number {
SUNXI_GPIO_A_START = 0,
SUNXI_GPIO_B_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_A), //32
SUNXI_GPIO_C_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_B), //64
SUNXI_GPIO_D_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_C), //96
SUNXI_GPIO_E_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_D), //128
SUNXI_GPIO_F_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_E), //160
SUNXI_GPIO_G_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_F), //192
SUNXI_GPIO_H_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_G), //224
SUNXI_GPIO_I_START = SUNXI_GPIO_NEXT(SUNXI_GPIO_H) //256
};
/* SUNXI GPIO number definitions */
#define SUNXI_GPA(_nr) (SUNXI_GPIO_A_START + (_nr))
#define SUNXI_GPB(_nr) (SUNXI_GPIO_B_START + (_nr))
#define SUNXI_GPC(_nr) (SUNXI_GPIO_C_START + (_nr))
#define SUNXI_GPD(_nr) (SUNXI_GPIO_D_START + (_nr))
#define SUNXI_GPE(_nr) (SUNXI_GPIO_E_START + (_nr))
#define SUNXI_GPF(_nr) (SUNXI_GPIO_F_START + (_nr))
#define SUNXI_GPG(_nr) (SUNXI_GPIO_G_START + (_nr))
#define SUNXI_GPH(_nr) (SUNXI_GPIO_H_START + (_nr))
#define SUNXI_GPI(_nr) (SUNXI_GPIO_I_START + (_nr))
/* GPIO pin function config */
#define SUNXI_GPIO_INPUT (0)
#define SUNXI_GPIO_OUTPUT (1)
#define SUNXI_GPIO_PER (2)
#define SUNXI_GPA0_ERXD3 (2)
#define SUNXI_GPA0_SPI1_CS0 (3)
#define SUNXI_GPA0_UART2_RTS (4)
#define SUNXI_GPA1_ERXD2 (2)
#define SUNXI_GPA1_SPI1_CLK	(3)
#define SUNXI_GPA1_UART2_CTS	(4)
#define SUNXI_GPA2_ERXD1	(2)
#define SUNXI_GPA2_SPI1_MOSI	(3)
#define SUNXI_GPA2_UART2_TX	(4)
#define SUNXI_GPA10_UART1_TX	(4)
#define SUNXI_GPA11_UART1_RX	(4)
#define SUN4I_GPB22_UART0_TX	(2)
#define SUN4I_GPB23_UART0_RX	(2)
#define SUN5I_GPG3_UART0_TX	(4)
#define SUN5I_GPG4_UART0_RX	(4)
#define SUNXI_GPC2_NCLE	(2)
#define SUNXI_GPC2_SPI0_CLK	(3)
#define SUNXI_GPC6_NRB0	(2)
#define SUNXI_GPC6_SDC2_CMD	(3)
#define SUNXI_GPC7_NRB1	(2)
#define SUNXI_GPC7_SDC2_CLK	(3)
#define SUNXI_GPC8_NDQ0	(2)
#define SUNXI_GPC8_SDC2_D0	(3)
#define SUNXI_GPC9_NDQ1	(2)
#define SUNXI_GPC9_SDC2_D1	(3)
#define SUNXI_GPC10_NDQ2	(2)
#define SUNXI_GPC10_SDC2_D2	(3)
#define SUNXI_GPC11_NDQ3	(2)
#define SUNXI_GPC11_SDC2_D3	(3)
#define SUNXI_GPF2_SDC0_CLK	(2)
#define SUNXI_GPF2_UART0_TX	(4)
#define SUNXI_GPF4_SDC0_D3	(2)
#define SUNXI_GPF4_UART0_RX	(4)
extern int sunxi_gpio_input(unsigned int pin);
extern int sunxi_gpio_init(void);
extern int sunxi_gpio_set_cfgpin(unsigned int pin, unsigned int val);
extern int sunxi_gpio_get_cfgpin(unsigned int pin);
extern int sunxi_gpio_output(unsigned int pin, unsigned int val);
extern void sunxi_gpio_cleanup(void);
extern unsigned int SUNXI_PIO_BASE;
#endif

src/main.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*-  */
/*
 * main.c
 * Copyright (C) 2015 Andrew Ivanov <search_terminal@mail.ru>
 * 
 * DHT is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * DHT is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//##define BCM2708_PERI_BASE        0x20000000
//#define GPIO_BASE                (BCM2708_PERI_BASE + 0x200000) /* GPIO controller */

//################################################################################​###########
//########################### Geändert zur verwendung auf dem Cubietruck
//################################################################################​##########
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <dirent.h>
#include <fcntl.h>
#include <assert.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include "gpio_lib.h"
#define MAXTIMINGS 500 //100 Задает значение тайминга
#define PG3  SUNXI_GPG(3)
int dhtpin = PG3;//##  Вывод к которому подключен датчик(в .fex файле)

//#define DEBUG

#define DHT11 11
#define DHT22 22
#define AM2302 22

int readDHT(int type, int pin);
int main(int argc, char **argv)

{
if(SETUP_OK!=sunxi_gpio_init()){

        printf("Ошибка инициализации!\n");

return -1;

}

  if (argc != 2) {

    printf("Укажите модель датчика: %s [11|22|2302]#\n", argv[0]);

    //printf("example: %s 2302 4 - Read from an AM2302 connected to GPIO #4\n", argv[0]);

    return 2;

  }

  int type = 0;
  if (strcmp(argv[1], "11") == 0) type = DHT11;
  if (strcmp(argv[1], "22") == 0) type = DHT22;
  if (strcmp(argv[1], "2302") == 0) type = AM2302;
  if (type == 0) {
    printf("Select 11, 22, 2302 as type!\n");
    return 3;
  }
//atoi(argv[2]);
// if (dhtpin <= 0) {
//    printf("Please select a valid GPIO pin #\n");
//    return 3;
//  }
  readDHT(type, dhtpin);
  return 0;
} // main

  int bits[250], data[100];
  int bitidx = 0;
  int readDHT(int type, int pin) {
  int counter = 0;
  int laststate = HIGH;
  int j=0;


// Переключаем вывод в режим выхода
  sunxi_gpio_set_cfgpin(pin,OUTPUT);
  sunxi_gpio_output(pin,HIGH);
  usleep(500000);//500000
  sunxi_gpio_output(pin,LOW);
  usleep(19000);//18000
  sunxi_gpio_output(pin,HIGH);
  usleep(5);//18
  sunxi_gpio_set_cfgpin(pin,INPUT);
  data[0] = data[1] = data[2] = data[3] = data[4] = 0;
	
//////////////////////////////////////////////////////////////////////////////
  int block;///Счетчик попыток ожидания нуля./////////////////////////////////
  block = 0;///Без этой части, если датчик не успевал вообще//////////////////
  /////////////сработать, или срабатывал быстрее чем мы начинаем//////////////
  /////////////ждать нуля, то так как на датчике в режиме/////////////////////
  /////////////ожидания всегда единица, цикл становился бесконечным.//////////
//////////////////////////////////////////////////////////////////////////////
  while (sunxi_gpio_input(pin) == 1) {   //Если на выходе 1 то ожидаем 0
       printf("Ожидаем нуля от датчика \n" );
//////////////////////////////////////////////////////////////////////////////
  block++;//Блокировка от бесконечного цикла//////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
  usleep(1);//Засыпаем на микросекунду////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
  if (block ==10)/////////////////////////////////////////////////////////////
	break;////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
	}
// Читаем данные!
int i;
for (i=0; i< MAXTIMINGS; i++) {
    counter = 0;
    while (sunxi_gpio_input(pin) == laststate) {
    counter++;
//////////////////////////////////////////////////////////////////////////////
//usleep(1); Задержка в одну микросекунду приводит к//////////////////////////
// потере данных, не использовать!////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
	if (counter == 2000)//1000
      break;
   }
    laststate = sunxi_gpio_input(pin);
    if (counter == 2000) break; //1000
    bits[bitidx++] = counter;
    if ((i>3) && (i%2 == 0)) {
      // Кладем каждый бит в байт переменной для хранения
      data[j/8] <<= 1;
      if (counter > 200)//200
        data[j/8] |= 1;
      j++;
    }
}

#ifdef DEBUG
for ( i=3; i<bitidx; i+=2) {
    printf("bit %d: %d\n", i-3, bits[i]);
    printf("bit %d: %d (%d)\n", i-2, bits[i+1], bits[i+1] > 200);
}

#endif


/////////////////////////////////////
if (j<=39){ //Если бит меньше чем 39, то данные получены не полностью
printf("Данные не могут быть считаны, возможно компьютер сильно загружен \n");
printf("и интерфейс работы с датчиком не работает на должной скорости\n");
//return 1;  // и нет смысла дальнейших действий
wait(2000);
goto LOOP;
}
////////////////////////////////////

	
  printf("Data (%d): 0x%x 0x%x 0x%x 0x%x 0x%x\n", j, data[0], data[1], data[2], data[3], data[4]);

  if ((j >= 39) &&  // if ((j >= 39) &&
      (data[4] == ((data[0] + data[1] + data[2] + data[3]) & 0xFF)) ) {
     // Вау!
     if (type == DHT11)
    printf("Temp = %d *C, Hum = %d \%\n", data[2], data[0]);
     if (type == DHT22) {
    float f, h;
    h = data[0] * 256 + data[1];
    h /= 10;
    f = (data[2] & 0x7F)* 256 + data[3];
        f /= 10.0;
        if (data[2] & 0x80)  f *= -1;
    printf("Temp =  %%.1f *C, Hum = %%.1f \%%\n", f, h);
    }

    return 1;

  }
  return 0;
/////////////////////////////////////////////
LOOP:   {


	if(SETUP_OK!=sunxi_gpio_init()){

        printf("Ошибка инициализации!\n");

  return -1;

   }
	
  readDHT(type, dhtpin); 
	
  return 0;

}
/////////////////////////////////////////////
}

